### PR TITLE
BLEGattsAttrMD: read_perm and write_perm

### PR DESCRIPTION
**Pull request by dangu**
When trying to read a characteristic I was getting READ_NOT_PERMITTED. I would like to discuss how to add the functionality of BLE_GAP_CONN_SEC_MODE_SET_OPEN() and friends to pc_ble_driver_py, to control the read_perm and write_perm of BLEGattsAttrMD.

I have made some tests with adding this line to BLEGattsAttrMD:

    attr_md.read_perm = driver.ble_gap_conn_sec_mode_t()

After that, it is possible to assign the sm and lv properties:

     attr_md.read_perm.sm=1
     attr_md.read_perm.lv=1

I also started implementing a class BLEGapConnSecMode for the ble_gap_conn_sec_mode_t but now I need help. How do I proceed?

Edit: Ok, I think I start to understand the structure of this. What do you think about my suggested set_no_access() and set_open() methods in BLEGapConnSecMode? Is this the way to go? This allows for this kind of code:

        read_perm = BLEGapConnSecMode()
        read_perm.set_open()
        write_perm = BLEGapConnSecMode()
        write_perm.set_open()
        
        attr_md = BLEGattsAttrMD(read_perm=read_perm, write_perm=write_perm)
        attr = BLEGattsAttr(uuid=char_128_uuid, attr_md=attr_md,
                            max_len=max_len, value=value)
        self.adapter.driver.ble_gatts_characteristic_add(
                            serv_128_handle.handle,
                            char_md, attr, char_128_handles)

### CODE REVIEW

**Comment by bihanssen** [comment]
@dangu thanks for the PR. I think your proposed solution looks sound.

File: pc_ble_driver_py/ble_driver.py [hunk 1]
```diff
@@ -369,6 +369,63 @@ def to_c(self):
...
+        
+    def to_c(self):
+        sec_mode = driver.ble_gap_conn_sec_mode_t()
+        sec_mode.sm = self.sm
+        sec_mode.lv = self.lv
+        
+        return sec_mode
+
+    def __str__(self):
+        return "sm({0.ssm}) lv({0.lv}))".format(
```
Comment (bihanssen) [inline pc_ble_driver_py/ble_driver.py:426]: Should be sm, not ssm?

File: pc_ble_driver_py/ble_driver.py [hunk 2]
```diff
@@ -369,6 +369,63 @@ def to_c(self):
...
+        
+    def set_lesc_enc_with_mitm(self): 
+        """BLE_GAP_CONN_SEC_MODE_SET_LESC_ENC_WITH_MITM"""
+        self.sm = 1
+        self.lv = 4
+    
+    def set_signed_no_mitm(self):
+        """BLE_GAP_CONN_SEC_MODE_SET_SIGNED_NO_MITM"""
+        self.sm = 2
+        self.lv = 1
```
Comment (bihanssen) [inline pc_ble_driver_py/ble_driver.py:405]: Missing newline

File: pc_ble_driver_py/ble_driver.py (hunk 1 above)
Comment (dangu) [reply pc_ble_driver_py/ble_driver.py:426]: Great catch! Yes, should be sm. How do we fix this? Should I push a fix?

**Comment by bihanssen** [comment]
@dangu, merging master into this branch is no problem. It’s anyway required before eventually merging into master.

**Review by kenr — APPROVED** [review]
Thanks!

**Comment by bihanssen** [comment]
@dangu I added some changes to the PR. Will merge this to master now.

**Pull request merged**